Editor - Add Video

aforcier opened the pull request:
Addresses https://github.com/wordpress-mobile/WordPress-Editor-Android/issues/274.

![274-video-upload-demo](https://cloud.githubusercontent.com/assets/9613966/13197445/891db196-d7bb-11e5-9b6f-d19a5cff3cc2.png)

* Adds support for adding video from WP media library (already uploaded)
* Adds support for uploading video to .org and .com sites (VideoPress required on .com)
* Just like [image uploads](https://github.com/wordpress-mobile/WordPress-Android/pull/3303#issuecomment-150593812), videos can be cancelled while uploading, retried if they fail, are automatically cancelled if the user closes the editor, and certain actions are disabled while uploads are in progress (HTML mode, publishing, opening the post settings dialog)
* Supports creating a post by sharing video from another app to WordPress
* Re-uses the compatibility upload overlays introduced to improve the API<19 upload experience ([ref](https://github.com/wordpress-mobile/WordPress-Editor-Android/pull/270)]

<h4>Notes</h4>
* Video posters (thumbnails) aren't automatically displayed when the video upload completes, as the video takes some time to process on the VideoPress side (for .org, there are no thumbnails available in the first place). The poster displays automatically once processing is complete and the thumbnail is made available on the server.
* In the future we might want to look into generating a thumbnail for local videos and using that as the placeholder while the video uploads. In the interest of time I omitted this from this PR as in my opinion the default placeholder image we're using (see above screenshot) is good enough for now.
* Some cleanup/refactoring of shared CSS and JS methods between images and video is called for, but I think that's something we can visit [in the future](https://github.com/wordpress-mobile/WordPress-Editor-Android/issues/287). For now I think it's better not to change anything unnecessary, to simplify diffing with iOS's ZSSEditor as we continue to improve both editors.

cc @maxme

maxme commented:
I quickly tested video upload, my first test was with a small video on a very good wifi connection. I noticed that the progress bar was at 0% (ie. hidden) for at least 20 seconds, it felt like my upload was stuck. I think we should add an animated loading indicator (animated progress bar?) or at least start the progress bar at 10% - this is not blocking, we can open a new issue for that.

Note: I've encountered several 500 error on a .org site, it's not related to the patch but we should maybe investigate (check the apache logs) and open a .org ticket.

maxme commented:
> at least start the progress bar at 10% - this is not blocking, we can open a new issue for that.

I've found a great way to fix this, the "optimistic progress bar" (same kind of thing is implemented in the Google Search app).  The progress bar starts to update slowly from 0 to 15% (and it does more and more slowly) even if the upload hasn't started.

maxme commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/EditorFragment.java:
```diff
@@ -744,12 +745,30 @@ public void appendMediaFile(final MediaFile mediaFile, final String mediaUrl, Im
+                        String videoPressId = ShortcodeUtils.getVideoPressIdFromShortCode(
+                                mediaFile.getVideoPressShortCode());
+
+                        mWebView.execJavaScriptFromString("ZSSEditor.insertVideo('" + mediaUrl + "', '" +
```
I wonder if the `mediaUrl` could be abused here (with infos coming from the server) and if we should escape it (same for the `ZSSEditor.insertImage`).

maxme commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/EditorFragment.java:
```diff
@@ -809,27 +828,48 @@ public void setContentPlaceholder(CharSequence placeholderText) {
+            mWebView.post(new Runnable() {
+                @Override
+                public void run() {
+                    String progressString = String.format(Locale.US, "%.1f", progress);
```
:+1:

maxme commented on libs/editor/libs/editor-common/assets/ZSSRichTextEditor.js:
```diff
@@ -1087,8 +1087,11 @@ ZSSEditor.markAllUploadingImagesAsFailed = function(message) {
     for(var i = 0; i < matches.size(); i++) {
         var mediaId = matches[i].getAttribute("data-wpid");
+        var videoId = matches[i].getAttribute("data-video_wpid");
         if (mediaId != null) {
```
From the [getAttribute](https://developer.mozilla.org/en-US/docs/Web/API/Element/getAttribute) documentation:

> If the given attribute does not exist, the value returned will either be null or "" (the empty string);

Considering there are differences on Android Webviews, I wonder if we should check with `hasAttribute` first or test if the returned value is the empty string.

aforcier replied:
That's a good point. I'm leaning towards checking for empty string instead of using `hasAttribute`: the less DOM methods we can rely on the better, for the same reason of Android WebView differences (by API and by OEM).

aforcier replied:
Nevermind, we already use `hasAttribute` after all :)

maxme commented on libs/editor/libs/editor-common/assets/ZSSRichTextEditor.js:
```diff
@@ -1138,45 +1144,76 @@ ZSSEditor.removeImage = function(imageNodeIdentifier) {
+ZSSEditor.insertVideo = function(videoURL, posterURL, videopressID) {
+    var html = '<video webkit-playsinline src="' + videoURL + '" onclick="" controls="controls" preload="metadata"';
+
+    if (posterURL != '') {
```
In this call: 
```
mWebView.execJavaScriptFromString("ZSSEditor.insertVideo('" + mediaUrl + "', '" +
    mediaFile.getThumbnailURL() + "', '" + videoPressId +  "');");
```

`mediaFile.getThumbnailURL() ` can be null, so `posterURL != null` is true, and probably showing a broken thumbnail  instead of a placeholder.

BTW, should we add a default placeholder if `posterURL == '' || posterURL == null)`?

aforcier replied:
> BTW, should we add a default placeholder if `posterURL == '' || posterURL == null`)?

We can't, there's an issue with video tags on Android where the video will stretch to fit the size of the placeholder image, even if the video has different dimensions (first encountered https://github.com/wordpress-mobile/WordPress-Android/commit/66bea0be6243139e40c34d71c7c5d3108d39b9a2). So we only show the poster while loading, and only use a thumbnail specific to that video (if it exists) once the video element has been updated with the remote `src` of the video.

Agreed on null-checking `posterURL` though.

maxme replied:
:cry: :panda_face:

maxme commented on libs/editor/libs/editor-common/assets/ZSSRichTextEditor.js:
```diff
@@ -1138,45 +1144,76 @@ ZSSEditor.removeImage = function(imageNodeIdentifier) {
-    this.insertHTML(html);
+
+    if (posterURL == '') {
+       posterURL = "wpposter.svg";
```
:+1: `posterURL` is never null here (thanks to a call to `StringUtils.notNullStr` on the Java side)

aforcier commented:
* Added null checking for `getThumbnailURL` calls
* Used `hasAttribute` to null-check when parsing for failed media items in the ZSSEditor
* Added quote escaping to all situations where server-obtained URLs are being passed to the ZSSEditor
* Opened a new issue for adding support for an 'optimistic progress bar': https://github.com/wordpress-mobile/WordPress-Editor-Android/issues/290

maxme commented:
:shipit:
